scenario_player/utils/reclaim.py

```diff
@@ -0,0 +1,520 @@
+    TokenAmount,
+    TokenNetworkAddress,
+    TokenNetworkRegistryAddress,
+    WithdrawAmount,
+)
+from scenario_player.utils.contracts import (
+    get_proxy_manager,
+    get_udc_and_corresponding_token_from_dependencies,
+)
+from scenario_player.utils.legacy import wait_for_txs
+
+log = structlog.get_logger(__name__)
+
+VALUE_TX_GAS_COST = 21_000
+
+
+@dataclass
+class ReclamationCandidate:
+    address: ChecksumAddress
+    keyfile_content: dict
+    node_dir: pathlib.Path
+
+    @cached_property
+    def privkey(self):
+        return decode_keyfile_json(self.keyfile_content, b"")
+
+
+def get_reclamation_candidates(
+    data_path: pathlib.Path, min_age_hours: int
+) -> List[ReclamationCandidate]:
+    candidates: List[ReclamationCandidate] = []
+    for node_dir in iter_chain(data_path.glob("**/node_???"), data_path.glob("**/node_*_???")):
+        if (node_dir / "reclaimed").exists():
+            continue
+
+        last_run = next(
+            iter(
+                sorted(
+                    list(node_dir.glob("run-*.log*")),
+                    key=lambda p: p.stat().st_mtime,
+                    reverse=True,
+                )
+            ),
+            None,
+        )
+        # If there is no last run assume we can reclaim
+        if last_run:
+            age_hours = (time.time() - last_run.stat().st_mtime) / 3600
+            if age_hours < min_age_hours:
+                scenario_name: Path = Path(node_dir.parent.name)
+                log.debug(
+                    "Skipping too recent node",
+                    scenario_name=scenario_name,
+                    node=node_dir.name,
+                    age_hours=age_hours,
+                )
+                continue
+        for keyfile in node_dir.glob("keys/*"):
+            keyfile_content = json.loads(keyfile.read_text())
+            address = keyfile_content.get("address")
+            if address:
+                candidates.append(
+                    ReclamationCandidate(
+                        address=to_checksum_address(address),
+                        node_dir=node_dir,
+                        keyfile_content=keyfile_content,
+                    )
+                )
+    return candidates
+
+
+def withdraw_from_udc(
+    reclamation_candidates: List[ReclamationCandidate],
+    contract_manager: ContractManager,
+    account: Account,
+    eth_rpc_endpoint: URI,
+):
+    web3 = Web3(HTTPProvider(eth_rpc_endpoint))
+    chain_id = ChainID(web3.eth.chainId)
+    deploy = get_contracts_deployment_info(chain_id, RAIDEN_CONTRACT_VERSION)
+    assert deploy
+
+    address_to_proxy_manager: Dict[ChecksumAddress, ProxyManager] = dict()
+    planned_withdraws: Dict[ChecksumAddress, Tuple[BlockNumber, TokenAmount]] = dict()
+
+    log.info("Checking chain for deposits in UserDeposit contact")
+    for node in reclamation_candidates:
+        if node.address not in address_to_proxy_manager:
+            client = JSONRPCClient(web3, node.privkey)
```

Comment:
Hummm ... I didn't see these before, I think this code was just moved instead of added to the PR so I didn't review it before. This has to be fixed otherwise we will most likely have `nonce` problems, only one instance of a `JSONRPCClient` can exist at any time because there is no external coordination.

Even if this instances don't leave the function this should be fixed.